designs/in-place-upgrades-eksa.md

```diff
@@ -79,14 +83,51 @@ There are a few ways to go about this:
 More thought is required in this area so it will require a follow up.
 
 ### Upgrading MachineDeployments
-We will have a `WorkersKubeadmUpgrade` CRD and implement a controller to reconcile it. This controller will be responsible for orchestrating the upgrade of the worker nodes: controlling the node sequence, define the upgrade steps required for each node and updating the CAPI objects (`Machine`, `KubeadmConfig`, etc.) after each node is upgraded.
-- The controller will upgrade worker nodes in the same `WorkersKubeadmUpgrade` one by one.
-- The upgrade actions will be defined as container specs that will be passed to the `NodeUpgrade` to execute and track.
+We will have a `MachineDeploymentUpgrade` CRD and implement a controller to reconcile it. This controller will be responsible for orchestrating the upgrade of the worker nodes: controlling the node sequence, define the upgrade steps required for each node and updating the CAPI objects (`Machine`, `KubeadmConfig`, etc.) after each node is upgraded.
+- The controller will upgrade worker nodes in the same `MachineDeployment` one by one.
+- The controller will create `NodeUpgrade` objects which will go and upgrade the nodes.
+
+This `MachineDeploymentUpgrade` should contain reference to the `MachineDeployment` that needs to be upgraded and a status that allows to track the progress of the upgrade.
+
+```go
+type MachineDeploymentUpgradeSpec struct {
+	MachineDeployment corev1.ObjectReference `json:"machineDeployment"`
+}
 
-This `WorkersKubeadmUpgrade` should contain information about the new component versions that will be installed in the nodes and a status that allows to track the progress of the upgrade.
+// MachineDeploymentUpgradeStatus defines the observed state of MachineDeploymentUpgrade.
+type MachineDeploymentUpgradeStatus struct {
+	RequireUpgrade int64          `json:"requireUpgrade,omitempty"`
+	Upgraded       int64          `json:"upgraded,omitempty"`
+	Ready          bool           `json:"ready,omitempty"`
+	MachineState   []MachineState `json:"machineState,omitempty"`
+}
+```
 
 #### Upgrading nodes
-We will have `NodeKubeadmUpgrade` CRD and implement a controller to reconcile it. This controller will be responsible from scheduling a pod on the specified workload cluster node with the specified containers as `initContainers`. It will track their progress and bubble up any error/success to the `NodeKubeadmUpgrade` status. The status should also allow to track the progress of the different upgrade "steps".
+We will have `NodeUpgrade` CRD and implement a controller to reconcile it. This controller will be responsible from scheduling a pod on the specified workload cluster node with `initContainers` with different commands depending on whether the node is a control plane node or a worker node. It will track their progress and bubble up any error/success to the `NodeUpgrade` status.
```

Comment:
```suggestion
We will have `NodeUpgrade` CRD and implement a controller to reconcile it. This controller will be responsible from scheduling a pod on the specified workload cluster node with `initContainers` with different commands to the upgrader pod script depending on whether the node is a control plane node or a worker node. It will track their progress and bubble up any error/success to the `NodeUpgrade` status.
```